### DISCUSSION TITLE
ESP-WROOM-32 doesn't wake from lightsleep on timer

### DISCUSSION THREAD

**Discussion by kebwi** [micropython/micropython#10097] [Hardware & Peripherals]
Light sleep is pretty straightforward on my D1 Minis. When I tried similar code on an ESP-WROOM-32, the device never woke up. All I'm doing is calling this:

`machine.lightsleep(sleep_duration_ms)  # e.g., 1000 to sleep for one second`

Just in case there some confusion about units and passing large values was actually indicating seconds instead of milliseconds, I've also tried calling it with very small numbers, such as 1. I never get the terminal prompt back, nor does the device reboot or otherwise rerun main from the top, as would be expected from deepsleep, which I'm not testing anyway. It simply hangs after calling lightsleep as shown above and must be reset with a physical push of the onboard reset button.

Do I need to tie some wake pin to some reset pin, as is necessary on some ESP8266 systems. Although that was only necessary for deepsleep anyway, not lightsleep. The ESP32 doesn't have clearly labeled wake or reset pins, although I'm practically certain the EN pin is the reset pin, but I still don't know which pin would be the wake pin. Some looking around suggest I might be able to configure any RTC GPIO as a wake pin tied to the EN pin -- but I still don't see why that would be necessary to merely make lightsleep work effectively? Should it just work directly from a single line of code, as above?

Thanks.

**Comment by robert-hh** [comment]
Just to complete the image, although it might not matter in you case: which firmware version do you use?
I just tried machine.lightsleep() with a WROOM equivalent board, Firmware 1.19.1-443, and it worked as expected,

**Reply by davefes** [reply]
>I never get the terminal prompt back, nor does the device reboot or otherwise rerun main from the top, as would be expected from deepsleep, which I'm not testing anyway. 

lightsleep just picks-up where it left off.  So, it doesn't reboot or re-run main.  I have two ESP32-WROOM boards running 1.18 and I think rshell does come to life and show you that things have started again after a lightsleep.  I find it very helpful to have a "heartbeat" LED in the main loop.

**Reply by kebwi** [reply]
Mine doesn't wake up. I'll keep fiddling with it. Thanks.

**Comment by fivdi** [comment]
Are you by any chance using the WebREPL? If so, calling `machine.lightsleep()` from the browser window will result in the browser loosing the connection to the ESP32 and it will look like everything is broken.

Calling `machine.lightsleep()` from the REPL via the USB port works for me.

If you are not using the WebREPL I would suggest posting a complete program that can be used to reproduce the problem. Please post the complete content of boot.py too. A link to the type of board being used might help too.

**Reply by kebwi** [reply]
I had posted the other day about trouble with the webREPL and learned in the process that I shouldn't expect the webREPL to come back. My confusion originated from the fact that the webREPL comes back just fine on an ESP8266. While I am still using the webREPL for monitoring, that is not the problem I am bringing up in this thread. My script does other things besides initiate a webREPL and those things don't happen. It should raise a GPIO to turn on some sensors. That would cause LED power illuminators on the sensors to come on, giving a visual indication of waking up, but that's difficult to confirm since the script only wakes up periodically, so sitting there staring at it waiting for the sensor lights to blink is tedious (I need to run a test with a short sleep interval).

Furthermore, the script should issue an http_get request to submit the sensor data to my web server and the web server is not receiving any such data submission. So that initially gave me a further sense that the device wasn't waking up.

I do think I isolated the problem though. The network connection is lost during light sleep. If I reconnect upon waking up, I do get data submissions. So lightsleep is working, but the device doesn't automatically reconnect to the network upon waking up. Again, my prior work with ESP8266s did not exhibit this problem. They remember the network configuration even across deep sleep and reconnect automatically. I have no idea why I have to reconnect manually with the ESP32.

Does it make sense to you that the ESP32 wouldn't reconnect to the network upon waking up?

**Reply by davefes** [reply]
https://docs.micropython.org/en/latest/library/esp.html
I am a bit confused as it talks about SLEEP_LIGHT not lightsleep().  However, I understood that the ESP32 did not remember the connection, so never tried otherwise.

There has been some recent discussion on whether or not the ESP32 will try to reconnect when it is brought back up.  karfas showed me that there is a WLAN.config(reconnect=N) available.

https://github.com/orgs/micropython/discussions/10029

Generally, I have the opposite problem and try to ensure that the ESP32 is NOT connected to anything when I re-boot.  To ensure that things always start up at square-one I do a poweron_reset()  see https://github.com/orgs/micropython/discussions/9345

**Reply by kebwi** [reply]
The function call is machine.lightsleep().

EDIT: Perhaps, umm, were you suggesting I'm using the wrong function to put the device to sleep? Were you recommending a different approach? Thanks.

**Reply by kebwi** [reply]
Thanks for the ref to WLAN.config(reconnect=N).

**Reply by davefes** [reply]
No, just didn't put machine in front of it.

**Reply by davefes** [reply]
In my watering timers I do:

```py
while True:
    make STA active true
    connect to it

    do my processing

    disconnect from it
    make STA active false

    lightsleep(x)
```

Part of my connect function is always to make sure that the ESP32 is disconnected:
```py
    if sta_if.isconnected():
        sta_if.disconnect()
        print ('started in the connected state')
        print ('but now it is disconnected')
    else:
        print ('started in the disconnected state')
```
